generate/templates/manual/repository/statistics.cc

```diff
@@ -0,0 +1,1889 @@
+public:
+  WorkItemOidStrType(void *objectInfo, git_object_t type)
+    : m_objectInfo(objectInfo), m_oid_type(type) {}
+  ~WorkItemOidStrType() = default;
+  WorkItemOidStrType(const WorkItemOidStrType &other) = delete;
+  WorkItemOidStrType(WorkItemOidStrType &&other) = delete;
+  WorkItemOidStrType& operator=(const WorkItemOidStrType &other) = delete;
+  WorkItemOidStrType& operator=(WorkItemOidStrType &&other) = delete;
+  
+  void* GetObjectInfo() const { return m_objectInfo; }
+  const git_object_t& GetOidType() const { return m_oid_type; }
+
+private:
+  void *m_objectInfo {nullptr};
+  git_object_t m_oid_type {};
+};
+
+/**
+ * \class WorkerReachCounter
+ * Worker for the WorkPool to count reachability of each object.
+ */
+class WorkerReachCounter : public IWorker
+{
+public:
+  WorkerReachCounter(OdbObjectsData *odbObjectsData)
+    : m_odbObjectsData(odbObjectsData) {}
+  ~WorkerReachCounter() = default;
+  WorkerReachCounter(const WorkerReachCounter &other) = delete;
+  WorkerReachCounter(WorkerReachCounter &&other) = delete;
+  WorkerReachCounter& operator=(const WorkerReachCounter &other) = delete;
+  WorkerReachCounter& operator=(WorkerReachCounter &&other) = delete;
+
+  bool Initialize() { return true; }
+  bool Execute(std::unique_ptr<WorkItem> &&work);
+
+private:
+  void setReachabilityFromTags(void *objectInfo);
+  void setReachabilityFromCommits(void *objectInfo);
+  void setReachabilityFromTrees(void *objectInfo);
+
+  OdbObjectsData *m_odbObjectsData {nullptr};
+};
+
+/**
+ * WorkerReachCounter::Execute
+ */
+bool WorkerReachCounter::Execute(std::unique_ptr<WorkItem> &&work)
+{
+  std::unique_ptr<WorkItemOidStrType> wi {static_cast<WorkItemOidStrType*>(work.release())};
+  void *objectInfo = wi->GetObjectInfo();
+  const git_object_t &oid_type = wi->GetOidType();
+
+  switch (oid_type) {
+    case GIT_OBJECT_TAG:
+      setReachabilityFromTags(objectInfo);
+      break;
+    case GIT_OBJECT_COMMIT:
+      setReachabilityFromCommits(objectInfo);
+      break;
+    case GIT_OBJECT_TREE:
+      setReachabilityFromTrees(objectInfo);
+      break;
+    case GIT_OBJECT_BLOB:
+      // do not process blobs in this stage
+      break;
+    default:
+      break;
+  }
+
+  return true;
+}
+
+/**
+ * WorkerReachCounter::setReachabilityFromTags
+ * Adds reachability counter where tags point (any type of object).
+ */
+void WorkerReachCounter::setReachabilityFromTags(void *objectInfo)
+{
+  const OdbObjectsData::TagInfo *tagInfo = static_cast<const OdbObjectsData::TagInfo *>(objectInfo);
+
+  switch (tagInfo->typeTarget) {
+    case GIT_OBJECT_COMMIT:
+    {
+      OdbObjectsData::iterCommitInfo itCommitInfo =
+      m_odbObjectsData->commits.info.find(tagInfo->oidTarget);
+    
+      if (itCommitInfo != m_odbObjectsData->commits.info.end()) {
+        { // lock
+          std::lock_guard<std::mutex> lock(m_odbObjectsData->infoMutex.commits);
+          ++itCommitInfo->second.reachability;
+        }
+      }
```

Comment:
There's a data race here: `m_odbObjectsData->commits.info` is read here without a lock, and possibly written to at the same time via `WorkerStoreOdbData::Execute`. Standard collections may be either read from multiple threads concurrently or written to by a single thread, but not both.

Reply:
Same with the tree and blob sections below.

Reply:
I've also added better comments to try to clarify the code.

To obtain the final result, the whole process is run in different stages:

>  // stage 1 methods: store data from repository (with threads)
>  int storeObjectsInfo();
>  int storeAndCountRefs();
>  // stage 2 methods: count reachability of each object (with threads)
>  // NOTE: we need this stage, since so far libgit2 doesn't provide unreachable objects
>  bool setObjectsReachability();
>  void setReachabilityFromRefs();
>  void setUnreachables();
>  // stage 3 methods: prune unreachable oids
>  void pruneUnreachables();
>  void pruneUnreachableTags();
>  void pruneUnreachableCommits();
>  void pruneUnreachableTrees();
>  void pruneUnreachableBlobs();
>  // stage 4 methods: repositorySize and biggestObjects
>  void statsCountAndMax();
>  // stage 5 methods: historyStructure and biggestCheckouts
>  bool statsHistoryAndBiggestCheckouts();
>  bool calculateBiggestCheckouts();
>  OdbObjectsData::iterTreeInfo calculateTreeStatistics(const std::string &oidTree);
>  bool calculateMaxTagDepth();
>  OdbObjectsData::iterTagInfo calculateTagDepth(const std::string &oidTag);

If a stage leverages threads via a worker pool, the worker pool is created and we wait until all the threads are done to continue with the next stage.

Stage 1 leveraged threads that run `WorkerStoreOdbData::Execute`, storing information in `m_odbObjectsData->commits.info`, `m_odbObjectsData->trees.info`, etc.

In this case at this point of the code we are in stage 2, where `WorkerReachCounter::Execute` is called instead of `WorkerStoreOdbData::Execute`, and threads from stage 1 have all stopped, hence there won't be a data race with the containers written during stage 1.

Thanks! since this needed a better explanation in the code 👍

Reply:
I see, yes that helps clarify :+1: